databuilder/main.py

```diff
@@ -14,6 +15,35 @@
 
 log = structlog.getLogger()
 
+DATABUILDER_CODE = True
+
+
+def is_databuilder(tb: traceback) -> bool:
+    """
+    Check if the code in the traceback frame is
+    from the Databuilder Code Base. This aims to remove
+    traceback involving other libaries such as importlib.
+    """
+    db_globals = tb.tb_frame.f_globals
+    return "DATABUILDER_CODE" in db_globals
+
+
+def user_code_traceback_level(tb: traceback, error_type: Exception) -> int:
```

Comment:
I'm not convinced this approach to filtering the stack is quite right.

At the point where we hit an exception we've got a kind of stack sandwich:
```mermaid
flowchart TD
  id1["databuilder code (main.py etc)"]
  -->
  id2["user code"]
  -->
  id3["databuilder code (query_langage.py etc)"]
```

And we want to top and tail this to just show the user code.

So I think this involves walking the whole stack and removing everything up to an including the `exec_module` call. And then keeping the rest of the stack up until the first frame from a module inside `databuilder/*`.

So the user doesn't see either our initial code which imported their module _or_ the internals of our query language code which they call into.

This is all a bit theoretical at the moment though. Until we start using this in practice and seeing what the errors actually look like it's hard to know what's going to be most helpful for users.

Reply:
Hi @evansd

Thanks for this. Would you be able to provide an example a dataset definition that would demonstrate this sandwich problem? @inglesp  and I have come up with a number of examples but they do not cause an error in the query_language for example and so do not demonstrate sandwich-stacks. Any ideas?

Reply:
I think this is the sort of thing I meant:
```py
from databuilder.query_language import Dataset, build_patient_table

table = build_patient_table("table", {"i": int, "b": bool})

dataset = Dataset()
dataset.v = table.i + table.b
```

If I import that I get a traceback like this:
```
Traceback (most recent call last):
  File "/usr/lib/python3.9/runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib/python3.9/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/home/dave/projects/ebmdatalab/databuilder/databuilder/__main__.py", line 180, in <module>
    main(sys.argv[1:], environ=os.environ)
  File "/home/dave/projects/ebmdatalab/databuilder/databuilder/__main__.py", line 40, in main
    dump_dataset_sql(
  File "/home/dave/projects/ebmdatalab/databuilder/databuilder/main.py", line 53, in dump_dataset_sql
    dataset_definition = load_definition(definition_file)
  File "/home/dave/projects/ebmdatalab/databuilder/databuilder/main.py", line 123, in load_definition
    module = load_module(definition_file)
  File "/home/dave/projects/ebmdatalab/databuilder/databuilder/main.py", line 144, in load_module
    spec.loader.exec_module(module)
  File "<frozen importlib._bootstrap_external>", line 850, in exec_module
  File "<frozen importlib._bootstrap>", line 228, in _call_with_frames_removed
  File "error_example.py", line 6, in <module>
    dataset.v = table.i + table.b
  File "/home/dave/projects/ebmdatalab/databuilder/databuilder/query_language.py", line 189, in __add__
    return _apply(qm.Function.Add, self, other)
  File "/home/dave/projects/ebmdatalab/databuilder/databuilder/query_language.py", line 335, in _apply
    qm_node = qm_cls(*qm_args)
  File "<string>", line 5, in __init__
  File "/home/dave/projects/ebmdatalab/databuilder/databuilder/query_model.py", line 103, in __post_init__
    validate_node(self)
  File "/home/dave/projects/ebmdatalab/databuilder/databuilder/query_model.py", line 378, in validate_node
    validate_types(node)
  File "/home/dave/projects/ebmdatalab/databuilder/databuilder/query_model.py", line 607, in validate_types
    raise TypeValidationError(
databuilder.query_model.TypeValidationError: Add.rhs requires 'databuilder.query_model.Series[int]' but got 'databuilder.query_model.Series[bool]'
In node:
Function.Add(lhs=SelectColumn(source=SelectPatientTable(name='table', schema=TableSchema(i=int, b=bool)), name='i'), rhs=SelectColumn(source=SelectPatientTable(name='table', schema=TableSchema(i=int, b=bool)), name='b'))
```

Leaving aside the fact that the `TypeValidationError` we generate is not very friendly, if we exclude all the `databuilder` and `importlib` frames we get something like:
```
  File "error_example.py", line 6, in <module>
    dataset.v = table.i + table.b
databuilder.query_model.TypeValidationError: Add.rhs requires 'databuilder.query_model.Series[int]' but got 'databuilder.query_model.Series[bool]'
In node:
Function.Add(lhs=SelectColumn(source=SelectPatientTable(name='table', schema=TableSchema(i=int, b=bool)), name='i'), rhs=SelectColumn(source=SelectPatientTable(name='table', schema=TableSchema(i=int, b=bool)), name='b'))
```

Which I think is a big improvement.